File: src/harness/tools/registry/SchemaValidator.ts
/**
 * Harness Layer 2: Tools - Schema 验证器
 *
 * JSON Schema 风格的工具参数验证
 * 防止参数注入和类型错误
 */

import { Logger } from '../../../utils/Logger';
import type { ToolParameterDef } from '../../types';

/** 验证结果 */
export interface SchemaValidationResult {
  valid: boolean;
  errors: string[];
  /** 自动修复后的参数 */
  sanitizedParams: Record<string, unknown>;
}

export class SchemaValidator {
  /**
   * 验证工具参数
   * @param params 用户传入的参数
   * @param parameterDefs 工具参数定义
   * @param requiredParams 必填参数名列表
   */
  validate(
    params: Record<string, unknown>,
    parameterDefs: Record<string, ToolParameterDef>,
    requiredParams: string[]
  ): SchemaValidationResult {
    const errors: string[] = [];
    const sanitizedParams: Record<string, unknown> = {};

    // 检查必填参数
    for (const required of requiredParams) {
      if (params[required] === undefined || params[required] === null) {
        errors.push(`缺少必填参数: ${required}`);
      }
    }

    // 验证每个参数
    for (const [paramName, paramDef] of Object.entries(parameterDefs)) {
      const value = params[paramName];

      // 参数不存在
      if (value === undefined || value === null) {
        // 使用默认值
        if (paramDef.default !== undefined) {
          sanitizedParams[paramName] = paramDef.default;
        }
        continue;
      }

      // 类型验证
      const typeError = this.validateType(paramName, value, paramDef);
      if (typeError) {
        errors.push(typeError);
        continue;
      }

      // 枚举验证
      if (paramDef.enum && !paramDef.enum.includes(String(value))) {
        errors.push(
          `参数 ${paramName} 的值 "${value}" 不在允许范围内: [${paramDef.enum.join(', ')}]`
        );
        continue;
      }

      // 数组元素验证
      if (paramDef.type === 'array' && Array.isArray(value) && paramDef.items) {
        for (let i = 0; i < value.length; i++) {
          const itemError = this.validateType(
            `${paramName}[${i}]`,
            value[i],
            paramDef.items
          );
          if (itemError) {
            errors.push(itemError);
          }
        }
      }

      // 对象属性验证
      if (
        paramDef.type === 'object' &&
        typeof value === 'object' &&
        value !== null &&
        paramDef.properties
      ) {
        const objResult = this.validate(
          value as Record<string, unknown>,
          paramDef.properties,
          [] // 嵌套对象的必填检查由外层控制
        );
        if (!objResult.valid) {
          errors.push(...objResult.errors.map((e) => `${paramName}.${e}`));
        }
      }

      sanitizedParams[paramName] = value;
    }

    // 检查未知参数（仅警告，不阻止执行）
    for (const paramName of Object.keys(params)) {
      if (!(paramName in parameterDefs)) {
        Logger.debug(`未知参数: ${paramName}（已忽略）`, 'SchemaValidator');
        sanitizedParams[paramName] = params[paramName];
      }
    }

    return {
      valid: errors.length === 0,
      errors,
      sanitizedParams,
    };
  }

  /**
   * 验证参数类型
   */
  private validateType(
    paramName: string,
    value: unknown,
    paramDef: ToolParameterDef
  ): string | null {
    const actualType = this.getTypeOf(value);

    if (actualType !== paramDef.type) {
      // 宽松类型转换: string "123" → number
      if (paramDef.type === 'number' && typeof value === 'string') {
        const num = Number(value);
        if (!isNaN(num)) {
          return null; // 允许字符串数字
        }
      }

      // 宽松类型转换: string "true" → boolean
      if (paramDef.type === 'boolean' && typeof value === 'string') {
        if (value === 'true' || value === 'false') {
          return null;
        }
      }

      return `参数 ${paramName} 类型错误: 期望 ${paramDef.type}, 实际 ${actualType}`;
    }

    return null;
  }

  /**
   * 获取值的类型名
   */
  private getTypeOf(value: unknown): string {
    if (value === null) return 'null';
    if (value === undefined) return 'undefined';
    if (Array.isArray(value)) return 'array';
    return typeof value;
  }
}

File: python/agent/tools/schema_validator.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from agent.core.logger import StructuredLogger
from agent.core.logger import log_ignored
log = StructuredLogger("schema_validator")



@dataclass
class ToolParameterDef:
    """工具参数定义。

    Attributes:
        name: 参数名称。
        type: 参数类型（string/number/boolean/array/object）。
        description: 参数描述。
        required: 是否必填。
        enum: 允许的枚举值列表。
        default: 默认值。
        items: 数组元素类型定义（type为array时）。
        properties: 对象属性定义（type为object时）。
    """

    name: str = ""
    type: str = "string"
    description: str = ""
    required: bool = True
    enum: list[str] | None = None
    default: Any = None
    items: dict[str, Any] | None = None
    properties: dict[str, Any] | None = None


@dataclass
class SchemaValidationResult:
    """参数校验结果。

    Attributes:
        valid: 是否校验通过。
        errors: 错误信息列表。
        sanitized_params: 清洗后的参数（含默认值填充）。
    """

    valid: bool
    errors: list[str] = field(default_factory=list)
    sanitized_params: dict[str, Any] = field(default_factory=dict)


class SchemaValidator:
    """工具参数Schema校验器。

    校验工具调用参数是否符合定义的参数规格，支持类型检查、必填校验、
    枚举值校验、嵌套对象校验和数组元素校验。

    内置类型兼容性：string→number（自动转换），string→boolean（true/false自动识别）。

    Usage:
        validator = SchemaValidator()
        result = validator.validate({"query": "test"}, parameter_defs)
        if not result.valid:
            logger.info(result.errors)
    """

    def validate(
        self,
        params: dict[str, Any],
        parameter_defs: dict[str, ToolParameterDef],
        required_params: list[str] | None = None,
    ) -> SchemaValidationResult:
        errors: list[str] = []
        sanitized: dict[str, Any] = {}

        required = required_params or [k for k, v in parameter_defs.items() if v.required]

        for param_name in required:
            if param_name not in params or params[param_name] is None:
                errors.append(f"缺少必填参数: {param_name}")

        for param_name, param_def in parameter_defs.items():
            value = params.get(param_name)

            if value is None:
                if param_def.default is not None:
                    sanitized[param_name] = param_def.default
                continue

            type_error = self._validate_type(param_name, value, param_def)
            if type_error:
                errors.append(type_error)
                continue

            if param_def.enum and str(value) not in param_def.enum:
                errors.append(f"参数 {param_name} 的值 \"{value}\" 不在允许范围内: [{', '.join(param_def.enum)}]")
                continue

            if param_def.type == "array" and isinstance(value, list) and param_def.items:
                items_def = param_def.items
                for i, item in enumerate(value):
                    item_def = ToolParameterDef(
                        type=items_def.get("type", "string"),
                        description=items_def.get("description", ""),
                    )
                    item_error = self._validate_type(f"{param_name}[{i}]", item, item_def)
                    if item_error:
                        errors.append(item_error)

            if param_def.type == "object" and isinstance(value, dict) and param_def.properties:
                nested = param_def.properties
                nested_defs: dict[str, ToolParameterDef] = {}
                for k, v in nested.items():
                    if isinstance(v, dict):
                        nested_defs[k] = ToolParameterDef(
                            type=v.get("type", "string"),
                            description=v.get("description", ""),
                        )
                nested_result = self.validate(value, nested_defs, [])
                if not nested_result.valid:
                    for e in nested_result.errors:
                        errors.append(f"{param_name}.{e}")

            sanitized[param_name] = value

        for param_name in params:
            if param_name not in parameter_defs:
                log.debug("未知参数（已忽略）", param=param_name)
                sanitized[param_name] = params[param_name]

        return SchemaValidationResult(
            valid=len(errors) == 0,
            errors=errors,
            sanitized_params=sanitized,
        )

    @staticmethod
    def _validate_type(param_name: str, value: Any, param_def: ToolParameterDef) -> str | None:
        actual_type = SchemaValidator._get_type_of(value)

        if actual_type != param_def.type:
            if param_def.type == "number" and isinstance(value, str):
                try:
                    float(value)
                    return None
                except ValueError as _exc:
                    log_ignored(log, "schema_validator.SchemaValidator._validate_type", _exc)

            if param_def.type == "boolean" and isinstance(value, str):
                if value in ("true", "false"):
                    return None

            return f"参数 {param_name} 类型错误: 期望 {param_def.type}, 实际 {actual_type}"

        return None

    @staticmethod
    def _get_type_of(value: Any) -> str:
        if value is None:
            return "null"
        if isinstance(value, bool):
            return "boolean"
        if isinstance(value, (int, float)):
            return "number"
        if isinstance(value, list):
            return "array"
        if isinstance(value, dict):
            return "object"
        return "string"